### PR TITLE
Memory saves and deletes. Cache dependant commits.

### PR DESCRIPTION
**Pull request by Sriep**
Add modified saveVersion and deleteVersions that only change the in-memory tree and don't commit changes to the database.
Add a maxCahce to nodedb, so that data is only written to the disk if the max Cache size is exceeded.

### CODE REVIEW

File: nodedb.go [hunk 1]
```diff
@@ -342,13 +400,25 @@ func (ndb *nodeDB) uncacheNode(hash []byte) {
 	}
 }
 
+func (ndb *nodeDB) MaxChacheSizeExceeded() bool {
```
Comment (jaekwon) [inline nodedb.go:403]: cacheSize

**Comment by jaekwon** [comment]
Thank you for the contribution.  It's been a while since I looked at this code and the orphan system.  I wish there were more of a design document that explained the previous orphan system, the changes, and the new code with ideally diagrams and prose to make this easy to understand for future maintainers about versioning and orphaning.

File: nodedb.go [hunk 2]
```diff
@@ -193,14 +213,23 @@ func (ndb *nodeDB) DeleteVersion(version int64, checkLatestVersion bool) {
...
+
 // Saves orphaned nodes to disk under a special prefix.
 // version: the new version being saved.
 // orphans: the orphan nodes created since version-1
 func (ndb *nodeDB) SaveOrphans(version int64, orphans map[string]int64) {
 	ndb.mtx.Lock()
 	defer ndb.mtx.Unlock()
 
-	toVersion := ndb.getPreviousVersion(version)
+	// When saving a new version, the previous version will always be one less than the new version
```
Comment (jaekwon) [inline nodedb.go]: I don't think this is true?

Comment (jaekwon) [inline nodedb.go:231]: What happens when you save a version (H), delete it (H, so now the latest saved state is H-1), but then you save again?  Is it H+1 or H?

File: nodedb.go [hunk 3]
```diff
@@ -222,6 +251,10 @@ func (ndb *nodeDB) deleteOrphans(version int64) {
 	// Will be zero if there is no previous version.
 	predecessor := ndb.getPreviousVersion(version)
 
+	ndb.deleteOrphansWithPredecessor(version, predecessor, nil)
+}
+
+func (ndb *nodeDB) deleteOrphansWithPredecessor(version, predecessor int64, unsavedOrphans *map[string]int64) {
```
Comment (jaekwon) [inline nodedb.go:257]: what is the purpose of this function w/ the unsavedOrphans arg?  It never gets called?

**Comment by Sriep** [comment]
Still updating this idea. 
Databases can get bloated with large numbers of orphans, and disk database size can become the critical issue. At this point, orphans can move from being the solution to becomeing the problem. 

This suggests a new paradigm where you don't use orphan except possibly in the most recent blocks. Instead, after deleting old blocks, you can scan the database removing nodes that are not part of a tree anymore.

**Pull request closed without merging**